File: csharp/AoC2021/Day04/Day04/Program.cs
using Day04;

if (args.Length != 1)
{
    Console.Error.WriteLine("Usage: $0 <input file>");
    Environment.Exit(1);
}

var (numbers, boards) = Solution.ParseInput(File.ReadLines(args[0]));

var result1 = Solution.Play(boards, numbers);
if (result1 == null)
    Console.Error.WriteLine("No one won the game of squid bingo!");
else
    Console.WriteLine($"Board #{result1.Value.Idx} won with {result1.Value.LastNumber}, " +
                      $"giving it a score of {result1.Value.Score}, " +
                      $"so the final answer is {result1.Value.LastNumber * result1.Value.Score}.");

boards.ForEach(board => board.Reset());

var result2 = Solution.FindLosingBoard(boards, numbers);
if (result2 == null)
    Console.Error.WriteLine("No one won the game of squid bingo!");
else
    Console.WriteLine($"The losing board is #{result2.Value.Idx}, " +
                      $"which will win with {result2.Value.LastNumber}, " +
                      $"giving it a score of {result2.Value.Score}, " +
                      $"so the final answer is {result2.Value.LastNumber * result2.Value.Score}.");

File: fsharp/AoC2021/Day04/Day04/Program.fs
open System
open System.IO

let args = Environment.GetCommandLineArgs()

if args.Length <> 2 then
    failwith "Usage: $0 <INPUT FILE>"

let lines = File.ReadLines args.[1] |> List.ofSeq

let parse (lines: string list) =
    let numbersLine, boardLines =
        match lines with
        | numbers :: "" :: boards -> (numbers, boards)
        | _ -> failwith "Invalid input"

    let numbers =
        numbersLine.Split(',')
        |> Seq.map byte
        |> List.ofSeq

    let boards = boardLines |> Board.parseBoards

    (numbers, boards)

let numbers, boards = parse lines

match Solution.play boards numbers with
| Some score -> printfn $"The score of the winning board is {score}."
| _ -> eprintfn "No board will win with these numbers."

match Solution.findLosingBoard boards numbers with
| Some score -> printfn $"The score of the losing board is {score}."
| _ -> eprintfn "No board will win with these numbers."
